scope.go

```diff
@@ -0,0 +1,705 @@
+		s.rc.releaseMemory(int64(size))
+		return err
+	}
+
+	return nil
+}
+
+func (s *ResourceScope) reserveMemoryForConstraints(size int) error {
+	if s.owner != nil {
+		return s.owner.ReserveMemory(size)
+	}
+
+	var reserved int
+	var err error
+	for _, cst := range s.constraints {
+		if err = cst.ReserveMemoryForChild(int64(size)); err != nil {
+			break
+		}
+		reserved++
+	}
+
+	if err != nil {
+		// we failed because of a constraint; undo memory reservations
+		for _, cst := range s.constraints[:reserved] {
+			cst.ReleaseMemoryForChild(int64(size))
+		}
+	}
+
+	return err
+}
+
+func (s *ResourceScope) releaseMemoryForConstraints(size int) {
+	if s.owner != nil {
+		s.owner.ReleaseMemory(size)
+		return
+	}
+
+	for _, cst := range s.constraints {
+		cst.ReleaseMemoryForChild(int64(size))
+	}
+}
+
+func (s *ResourceScope) ReserveMemoryForChild(size int64) error {
+	s.Lock()
+	defer s.Unlock()
+
+	if s.done {
+		return ErrResourceScopeClosed
+	}
+
+	return s.rc.reserveMemory(size)
+}
+
+func (s *ResourceScope) ReleaseMemory(size int) {
+	s.Lock()
+	defer s.Unlock()
+
+	if s.done {
+		return
+	}
+
+	s.rc.releaseMemory(int64(size))
+	s.releaseMemoryForConstraints(size)
+}
+
+func (s *ResourceScope) ReleaseMemoryForChild(size int64) {
+	s.Lock()
+	defer s.Unlock()
+
+	if s.done {
+		return
+	}
+
+	s.rc.releaseMemory(size)
+}
+
+func (s *ResourceScope) AddStream(dir network.Direction) error {
+	s.Lock()
+	defer s.Unlock()
+
+	if s.done {
+		return ErrResourceScopeClosed
+	}
+
+	if err := s.rc.addStream(dir); err != nil {
+		return err
+	}
+
+	if err := s.addStreamForConstraints(dir); err != nil {
+		s.rc.removeStream(dir)
+		return err
+	}
+
+	return nil
+}
+
+func (s *ResourceScope) addStreamForConstraints(dir network.Direction) error {
+	if s.owner != nil {
+		return s.owner.AddStream(dir)
+	}
+
+	var err error
+	var reserved int
+	for _, cst := range s.constraints {
+		if err = cst.AddStreamForChild(dir); err != nil {
+			break
+		}
+		reserved++
+	}
+
+	if err != nil {
+		for _, cst := range s.constraints[:reserved] {
+			cst.RemoveStreamForChild(dir)
+		}
+	}
+
+	return err
+}
+
+func (s *ResourceScope) AddStreamForChild(dir network.Direction) error {
+	s.Lock()
+	defer s.Unlock()
+
+	if s.done {
+		return ErrResourceScopeClosed
+	}
+
+	return s.rc.addStream(dir)
+}
+
+func (s *ResourceScope) RemoveStream(dir network.Direction) {
+	s.Lock()
+	defer s.Unlock()
+
+	if s.done {
+		return
+	}
+
+	s.rc.removeStream(dir)
+	s.removeStreamForConstraints(dir)
+}
+
+func (s *ResourceScope) removeStreamForConstraints(dir network.Direction) {
+	if s.owner != nil {
+		s.owner.RemoveStream(dir)
+		return
+	}
+
+	for _, cst := range s.constraints {
+		cst.RemoveStreamForChild(dir)
+	}
+}
+
+func (s *ResourceScope) RemoveStreamForChild(dir network.Direction) {
+	s.Lock()
+	defer s.Unlock()
+
+	if s.done {
+		return
+	}
+
+	s.rc.removeStream(dir)
+}
+
+func (s *ResourceScope) AddConn(dir network.Direction) error {
```

Comment:
Can add a `bool` parameter here for `usesFD`, and absorb the functionality of `AddFD`?

Otherwise, when adding a (fd-consuming) connection, we have to call `AddConn` and `AddFD`, which requires locking the mutex twice.

Reply:
ok

Reply:
done.